connect/runtime/src/main/java/org/apache/kafka/connect/util/ConvertingFutureCallback.java

```diff
@@ -38,21 +44,33 @@ public ConvertingFutureCallback(Callback<T> underlying) {
 
     @Override
     public void onCompletion(Throwable error, U result) {
-        this.exception = error;
-        this.result = convert(result);
-        if (underlying != null)
-            underlying.onCompletion(error, this.result);
-        finishedLatch.countDown();
+        synchronized (this) {
+            if (isDone()) {
+                return;
+            }
+            this.exception = error;
+            this.result = convert(result);
+            if (underlying != null)
+                underlying.onCompletion(error, this.result);
+            finishedLatch.countDown();
+        }
     }
 
     @Override
     public boolean cancel(boolean b) {
-        return false;
+        synchronized (this) {
+            if (isDone()) {
+                return false;
+            }
+            this.cancelled = true;
+            finishedLatch.countDown();
+            return true;
```

Comment:
Do we have tests to test the cancellation logic?

Reply:
Added those now.